### DISCUSSION TITLE
Unable to sign a merged document

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#102] [Q&A] [answered]
I did a simple merger using pyPDF2 and used the same code which worked for a regular pdf. If you can help, it'd be reallly helpful.

`
root@gomu:/pdfsigner# python3 main.py 
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pyhanko/sign/fields.py", line 1303, in prepare_sig_field
    fields = form['/Fields']
  File "/usr/local/lib/python3.8/dist-packages/pyhanko/pdf_utils/generic.py", line 977, in __getitem__
    return dict.__getitem__(self, key).get_object()
KeyError: '/Fields'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 17, in <module>
    fields.append_signature_field(pdf_out=w,
  File "/usr/local/lib/python3.8/dist-packages/pyhanko/sign/fields.py", line 1472, in append_signature_field
    field_created, sig_field_ref = prepare_sig_field(
  File "/usr/local/lib/python3.8/dist-packages/pyhanko/sign/fields.py", line 1305, in prepare_sig_field
    raise ValueError('/AcroForm has no /Fields')
ValueError: /AcroForm has no /Fields`

Please help.

**Comment by MatthiasValvekens** [comment, marked as answer]
For some reason, the **AcroForm** dictionary in your input file doesn't contain a **Fields** entry. That's a file format error, and also very bizarre because **AcroForm** serves no purpose other than managing form fields.

The easiest way to work around this particular instance of the problem is to delete the AcroForm entry before letting pyHanko do its thing---assuming you know for sure that all your input files won't contain legitimate form data. Assuming you have the file open in an `IncrementalPdfFileWriter`, it should suffice to put in something like

```python
try:
    del my_writer.root['/AcroForm']
except KeyError:
   pass
```

Again, it's a workaround, but this is a pretty clear-cut case of bad input.

Hope that helps :)

**Reply by Vishwamithra37** [reply]
Hey,
So, I figured out the error. 

When I do `merge.append` in pyPDF2, it is messing up and removing all the form fields. I STRONGLY support a simple merge feature which does not mess with the Acroforms in pyHanko. It'll be PERFECTO. My 2 cents :) . As the pyHanko works after flattening/ (Using print to pdf in windows).

**Reply by MatthiasValvekens** [reply]
Glad that you managed to resolve your issue! :)


On the topic of adding merge support: I agree that merge support would be useful, but I try to make a conscious effort to keep pyHanko's feature set in scope. It was never my intention to turn pyHanko into a general-purpose PDF manipulation library (there are tons of those already!). PyHanko is supposed to handle PDF's cryptographic feature set. Other than that, it includes a few bells and whistles to generate pretty signatures, but that's more or less it.

A simple merge feature that discards all tagging, form data, bookmarks, etc. wouldn't be hard to implement, but I don't want to be on the hook for maintaining it as part of pyHanko. If I start putting in features of this type, it probably wouldn't be long before I'd start receiving bug reports along the lines of "hey, your merge function screwed up my bookmarks", etc. etc. And there's nothing wrong with that! If it's in the codebase, then by definition it's fair game for bug reports. As a maintainer, though, I don't want to deal with that _within pyHanko_, that's all.

TL;DR: There are not so many general-purpose PDF libraries with decent signing support. There are, however, quite a few PDF libraries that are capable of performing common tasks like merging/splitting/flattening/.... PyHanko was never meant to replace those, but rather to _complement_ them.

---------------


That said, pyHanko is free software: if someone wants to fork the project and put in merge/split features, or ship a separate library that uses pyHanko's internals to implement all that extra stuff, they're more than welcome to do just that. It's just that that person (probably) won't be me ;).
